### PR TITLE
(GH-104) Created base class for nuspec rules

### LINKED ISSUE #104: Create base class to help creating rules easier
## Expected Behavior
Creating new rules should be as easy as possible.

## Current Behavior
Currently we need to add all the logic our self, which results in duplication of code.

## Possible Solution
Create a base class with methods to easily create diagnostics, getting nuspec elements, etc.

## Additional
I'm already working on creating this base class.

### PR DESCRIPTION
**Pull request by AdmiringWorm**
## Description
With the changes in this PR we can now rely on a base class when creating rules for nuspec files.
The base class implements a few helper functions to create diagnostics for requirements/guidelines/suggestions and notes, as well as a helper to get a single element by its name.

Changes have also been made to the existing rules to make use of the new base class.

## Related Issue

fixes #104 

## Motivation and Context

To make it easier to implement rules.

## How Has This Been Tested?

Been tested locally.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality) *(I think)*
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/Chocolatey.Language.Server/Validations/DescriptionLengthValidation.cs [hunk 1]
```diff
@@ -13,53 +13,45 @@ namespace Chocolatey.Language.Server.Validations
...
             if (descriptionElement == null)
             {
-                yield return new Diagnostic {
-                    Message = "Description is required. See https://github.com/chocolatey/package-validator/wiki/DescriptionNotEmpty",
-                    Severity = DiagnosticSeverity.Error,
-                    Range = textPositions.GetRange(0, syntaxTree.End)
-                };
+                yield return CreateDiagnostic(
+                    0,
+                    syntaxTree.End,
```
Comment (mkevenaar) [inline src/Chocolatey.Language.Server/Validations/DescriptionLengthValidation.cs]: Why don't we use the start / end of the `<metadata>` element as the `<description>` should be 
there?

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Language.Xml;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using DiagnosticSeverity = OmniSharp.Extensions.LanguageServer.Protocol.Models.DiagnosticSeverity;
+
+namespace Chocolatey.Language.Server.Validations
```
Comment (mkevenaar) [inline src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: Just wondering if this class should be here or in the `Chocolatey.Language.Server` namespace. This is not actually an validation, but a blueprint.

@gep13 what do you think?

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        /// <returns>An enumerable of failed checks</returns>
+        public abstract IEnumerable<Diagnostic> Validate(XmlDocumentSyntax syntaxTree);
+
+        /// <summary>
+        /// Finds a single element in the specified <paramref name="syntaxTree"/> by the name.
+        /// </summary>
+        /// <param name="syntaxTree">The syntax tree to find the element in.</param>
+        /// <param name="name">The name of the element.</param>
+        /// <returns>The element syntax if one is found; otherwise returns <c>null</c>.</returns>
+        protected XmlElementSyntax FindElementByName(XmlDocumentSyntax syntaxTree, string name)
```
Comment (mkevenaar) [inline src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: We might want to create an `XmlElementSyntaxExtensions` class like we did for the `UriExtensions`

File: src/Chocolatey.Language.Server/Validations/DescriptionLengthValidation.cs (hunk 1 above)
Comment (AdmiringWorm) [reply src/Chocolatey.Language.Server/Validations/DescriptionLengthValidation.cs]: That could certainly be changed, didn't want to do it in this PR though as that wasn't the focus of it.

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs (hunk 2 above)
Comment (AdmiringWorm) [reply src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: IMO, the base class isn't a blueprint, and even if it was I would still suggest it to be placed in the Validations namespace as it's expected to be used by the rules only.

**Comment by gep13** [comment]
@AdmiringWorm I have merged this here: https://github.com/gep13/chocolatey-vscode/commit/71056dfc5d425a444285130009db98d8acd61fe8

**Comment by gep13** [comment]
@AdmiringWorm your changes have been merged, thanks for your contribution 👍

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs (hunk 3 above)
Comment (gep13) [reply src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: @mkevenaar this is a good idea as well.  We could do this as a separate scope of work.

**Pull request closed without merging**